crowbar_framework/app/models/api/crowbar.rb

```diff
@@ -80,6 +80,60 @@ def maintenance_updates_installed?
       end
     end
 
+    def compute_resources_available?
+      compute_nodes = NodeObject.find("roles:nova-compute-kvm")
+      return true if compute_nodes.empty?
+      if compute_nodes.size == 1
+        Rails.logger.warn("Only one compute node found; non-disruptive upgrade is not possible!")
+        return false
+      end
+
+      # Now, find out if we have enough memory for migrating
+
+      # Find the controller responsible for the nova services
+      founders = NodeObject.find(
+        "roles:nova-controller AND pacemaker_founder:true AND pacemaker_config_environment:*"
+      )
+      controller = founders.first
+
+      free_compute_memory = 0
+      largest_compute_node_total = 0
+      largest_compute_node_free = 0
+
+      compute_nodes.each do |compute_node|
+        # FIXME: are we sure nova uses the short name of node for its hosts?
+        name = compute_node.handle
+
+        total_cmd = ". /root/.openrc; /usr/bin/nova host-describe #{name} | grep '(total)'"
+        ssh_retval = controller.run_ssh_cmd(total_cmd)
+        out = ssh_retval[:stdout] || ""
+        total_memory = out.split[7].to_i
+
+        used_cmd = ". /root/.openrc; /usr/bin/nova host-describe #{name} | grep '(used_now)'"
+        ssh_retval = controller.run_ssh_cmd(used_cmd)
+        out = ssh_retval[:stdout] || ""
+        used_memory = out.split[7].to_i
+
+        free_memory = total_memory - used_memory
+        free_compute_memory += free_memory
+
+        # find the node with largest total memory
+        if total_memory > largest_compute_node_memory
+          largest_compute_node_total = total_memory
+          largest_compute_node_free = free_memory
+        end
+      end
+
+      if free_compute_memory - largest_compute_node_free < largest_compute_node_total
```

Comment:
@dguitarbite and @jsuchome, if I understand the motivation correctly then this is about making sure before the upgrade that the upgrade will not fail because of a lack of nova resources. We are assuming that we have to migrate an instance out before the upgrade of a node, and that we can't `virsh save` the instance on the node during the upgrade.

The problem I see with this approach here is that it will fail in certain situations. Let `N` be the number of nodes in the cluster and `m` the average number of free bytes per node. Then the total free memory is `N * m`. If `N` is large, `N * m` will also be large. Suppose that we want to migrate an instance that happens to use `M >> m` bytes. It is likely to not fit anywhere else. But by considering only something like `M < N * m` or similar we will conclude that we can proceed with the upgrade. I don't know whether such an unbalanced load is common, maybe this is really just a corner case. But it might be worth considering.